pkg/services/query/query.go

```diff
@@ -77,27 +85,71 @@ func (s *ServiceImpl) Run(ctx context.Context) error {
 }
 
 // QueryData processes queries and returns query responses. It handles queries to single or mixed datasources, as well as expressions.
-func (s *ServiceImpl) QueryData(ctx context.Context, user *user.SignedInUser, skipCache bool, reqDTO dtos.MetricRequest) (*backend.QueryDataResponse, error) {
+// By default, it will attempt to retrieve a cached response for the given query, unless `skipQueryCache` = true.
+// This is distinct from the `skipDSCache` argument, which only determines if the datasource metadata is cached.
+// Query caching is only implemented in Grafana Enterprise.
+func (s *ServiceImpl) QueryData(ctx context.Context, user *user.SignedInUser, skipDSCache bool, skipQueryCache bool, reqDTO dtos.MetricRequest) (*backend.QueryDataResponse, error) {
+	start := time.Now() // time how long this request takes
+
+	// First look in the query cache if enabled
+	var cr caching.CachedDataResponse
+	if !skipQueryCache {
+		if r := s.cachingService.HandleQueryRequest(ctx, reqDTO); r.Status == caching.StatusCacheHit {
+			caching.MarkCacheStatus(skipQueryCache, r)
+			return r.Response, nil
+		} else {
+			cr = r
+		}
+	}
+
+	// do the actual queries
+	resp, err := s.queryData(ctx, user, skipDSCache, reqDTO)
+
+	// Update the query cache with the result for this metrics request
+	if err == nil {
+		if cr.UpdateCacheFn != nil {
+			cr.UpdateCacheFn(ctx, resp)
+		}
+		// Update cache response headers
+		caching.MarkCacheStatus(skipQueryCache, cr)
+	}
+
+	duration := time.Since(start)
+
+	// record request duration
+	QueryRequestHistogram.With(prometheus.Labels{
+		"datasource_type": getDatasourceType(reqDTO.GetUniqueDatasourceTypes()),
+		"cache":           caching.GetCacheHeaders(resp)["X-Cache"],
+		"query_type":      getQueryType(user),
+	}).Observe(duration.Seconds())
+
+	return resp, err
+}
+
+// queryData contains the logic for processing and executing queries
+func (s *ServiceImpl) queryData(ctx context.Context, user *user.SignedInUser, skipDSCache bool, reqDTO dtos.MetricRequest) (*backend.QueryDataResponse, error) {
```

Comment:
This ends up being identical to the previous `QueryData` method - and `QueryData` handles the query caching, metrics, and calling this method.